flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/source/enumerator/splitter/MongoSampleSplitter.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * <p>Samples the collection to generate partitions.
+ *
+ * <p>Uses the average document size to split the collection into average sized chunks
+ *
+ * <p>The partitioner samples the collection, projects and sorts by the partition fields. Then uses
+ * every {@code samplesPerPartition} as the value to use to calculate the partition boundaries.
+ *
+ * <ul>
+ *   <li>scan.partition.size: The average size (MB) for each partition. Note: Uses the average
+ *       document size to determine the number of documents per partition so may not be even.
+ *       Defaults to: 64mb.
+ *   <li>scan.partition.samples: The number of samples to take per partition. Defaults to: 10. The
+ *       total number of samples taken is calculated as: {@code samples per partition * (count of
+ *       documents / number of documents per partition)}.
+ * </ul>
+ */
+@Internal
+public class MongoSampleSplitter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoSampleSplitter.class);
+
+    public static final MongoSampleSplitter INSTANCE = new MongoSampleSplitter();
+
+    private MongoSampleSplitter() {}
+
+    public Collection<MongoScanSourceSplit> split(MongoSplitContext splitContext) {
+        MongoReadOptions readOptions = splitContext.getReadOptions();
+        MongoNamespace namespace = splitContext.getMongoNamespace();
+
+        long count = splitContext.getCount();
+        long partitionSizeInBytes = readOptions.getPartitionSize().getBytes();
+        int samplesPerPartition = readOptions.getSamplesPerPartition();
+
+        long avgObjSizeInBytes = splitContext.getAvgObjSize();
+        if (avgObjSizeInBytes == 0L) {
+            LOG.info(
+                    "{} seems to be an empty collection, Returning a single partition.", namespace);
+            return MongoSingleSplitter.INSTANCE.split(splitContext);
+        }
+
+        long numDocumentsPerPartition = partitionSizeInBytes / avgObjSizeInBytes;
+        if (numDocumentsPerPartition >= count) {
+            LOG.info(
+                    "Fewer documents ({}) than the number of documents per partition ({}), Returning a single partition.",
+                    count,
+                    numDocumentsPerPartition);
+            return MongoSingleSplitter.INSTANCE.split(splitContext);
+        }
+
+        int numberOfSamples =
+                (int) Math.ceil((samplesPerPartition * count * 1.0d) / numDocumentsPerPartition);
+
+        List<BsonDocument> samples =
+                splitContext
+                        .getMongoCollection()
+                        .aggregate(
+                                Arrays.asList(
+                                        Aggregates.sample(numberOfSamples),
+                                        Aggregates.project(Projections.include(ID_FIELD)),
+                                        Aggregates.sort(Sorts.ascending(ID_FIELD))))
+                        .allowDiskUse(true)
+                        .into(new ArrayList<>());
```

Comment:
As an idea, add MIN_KEY/MAX_KEY also to this list. Just treat these 2 just like other samples, and ask for 2 fewer samples from MongoDB.

Unless there is a good reason to treat MIN/MAX as special boundaries. It was a bit odd to me that we always have 2 more partitions than we'd actually like. (`[MIN, sample #1]` and `[sample #N, MAX]`)

Reply:
Thanks @zentol for this great idea.
I have refactored the splitter using MIN to replace the first sample and MAX to replace the last sample.

Reply:
> I have refactored the splitter using MIN to replace the first sample and MAX to replace the last sample.

I'm not sure if this way is better than two fewer samples, but it's easier for coding. 
If this way is not very good, I will adjust it again.